### PR TITLE
add ECMWF eccodes

### PR DESCRIPTION
**Pull request by kmuehlbauer**
This adds the ECMWF eccodes library, the successor of grib_api and bufr_lib. This library is in beta status, but nevertheless usable.

### CODE REVIEW

File: recipes/eccodes/build.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+cd build
+cmake .. -DCMAKE_INSTALL_PREFIX=$PREFIX \
+         -DENABLE_JASPER=1 \
+         -DENABLE_NETCDF=1 \
+         -DENABLE_PNG=1 \
+         -DENABLE_PYTHON=0 \
+         -DENABLE_FORTRAN=0
+
+make
+ctest
```
Comment (ocefpaf) [inline recipes/eccodes/build.sh]: :+1:

File: recipes/eccodes/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,43 @@
+{% set version = "0.13.0" %}
+
+package:
+  name: eccodes
+  version: {{ version }}
+
+source:
+  url: https://software.ecmwf.int/wiki/download/attachments/45757960/eccodes-{{ version }}-Source.tar.gz?api=v2
```
Comment (ocefpaf) [inline recipes/eccodes/meta.yaml]: I don't think you need the `?api=v2` to download the source, do you?

File: recipes/eccodes/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  number: 0
+  skip: true  # [win]
+
+requirements:
+  build:
+    - cmake
+    - boost
+    - jasper
+    - libpng
+    - libnetcdf
```
Comment (ocefpaf) [inline recipes/eccodes/meta.yaml]: Can you change these to the pinned versions. (See https://github.com/conda-forge/staged-recipes/wiki/Pinned-dependencies)

```yaml
- libpng 1.6*
- libnetcdf 4.3*
```

You will also need to change them below.

File: recipes/eccodes/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    - ls $(codes_info -d)
+    
+about:
+  home: https://software.ecmwf.int/wiki/display/ECC/ecCodes+Home
+  license: Apache License Version 2.0, January 2004
+  summary: ECMWF ecCodes Copyright 2005-2016 ECMWF.
+
+extra:
+  recipe-maintainers:
+    - kmuehlbauer
```
Comment (ocefpaf) [inline recipes/eccodes/meta.yaml]: @pelson do you or anyone you know is interest in this recipe?

**Comment by kmuehlbauer** [comment]
ping @ocefpaf, @rhattersley, @pelson (maintainers of the conda-forge/ecmwf_grib-feedstock)

I followed the approach to first create the plain library and adding the bindings (python, fortran) as separate feedstock afterwards. Is this how we should do it? And, should I add you also as maintainers here?

**Comment by kmuehlbauer** [comment]
@ocefpaf do you know, if this is vital also here:
https://github.com/conda-forge/ecmwf_grib-feedstock/blob/master/recipe/meta.yaml#L15

Since eccodes has grib_api under the hood?

**Comment by ocefpaf** [comment]
> I followed the approach to first create the plain library and adding the bindings (python, fortran) as separate feedstock afterwards. Is this how we should do it? And, should I add you also as maintainers here?

I think so. I remember someone mentioning that `eccodes` is not py3k compat yet, so splitting will give us a c-lib that we can install on Python 3.

> @ocefpaf do you know, if this is vital also here:
https://github.com/conda-forge/ecmwf_grib-feedstock/blob/master/recipe/meta.yaml#L15

> Since eccodes has grib_api under the hood?

Sorry. I do not known if `detect_binary_files_with_prefix: true` is needed. Let's see if the tests passes without it and add it afterwards if things are broken.

File: recipes/eccodes/build.sh [hunk 5]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env bash
+
+mkdir build
+cd build
+cmake .. -DCMAKE_INSTALL_PREFIX=$PREFIX \
```
Comment (ocefpaf) [inline recipes/eccodes/build.sh]: Do you think we can compile that on Windows?

Comment (kmuehlbauer) [inline recipes/eccodes/build.sh]: I'm not sure. In the sources there is a "windows"-folder with .vcproj files for several grib/bufr programs. Shall we try?

What has to go into the bld.bat?

**Comment by kmuehlbauer** [comment]
linux is fine on CircleCI until the last test. Obviously we have to (re-)define those PATH-Values:

```
Default definition files path is used: /opt/conda/envs/_build/share/eccodes/definitions
Definition files path can be changed setting ECCODES_DEFINITION_PATH environment variable

Default SAMPLES path is used: /opt/conda/envs/_build/share/eccodes/samples
SAMPLES path can be changed setting ECCODES_SAMPLES_PATH environment variable
```

Do you have an idea where we should set that?

File: recipes/eccodes/build.sh (hunk 5 above)
Comment (ocefpaf) [inline recipes/eccodes/build.sh]: @kmuehlbauer I am not a Windows expert so let's leave it alone and open an issue once the `feedstock` is created. I usually ask @gillins for help :stuck_out_tongue_winking_eye:

**Comment by ocefpaf** [comment]
> linux is fine on CircleCI until the last test. Obviously we have to (re-)define those PATH-Values:

```
Default definition files path is used: /opt/conda/envs/_build/share/eccodes/definitions
Definition files path can be changed setting ECCODES_DEFINITION_PATH environment variable

Default SAMPLES path is used: /opt/conda/envs/_build/share/eccodes/samples
SAMPLES path can be changed setting ECCODES_SAMPLES_PATH environment variable
```

> Do you have an idea where we should set that? 

I am not familiar with `eccodes`, but it does look like we will need a solution similar to `gdal` to set that env var at `activation` time.

**Comment by kmuehlbauer** [comment]
> I am not familiar with eccodes, but it does look like we will need a solution similar to gdal to set that env var at activation time.

Seems that we could use something like that: http://conda.pydata.org/docs/building/meta-yaml.html?highlight=detect_binary_files_with_prefix#id27

If correctly installed the package should use the default path's and they should poitn to the conda installation, not?

**Comment by msarahan** [comment]
Regarding environment variable setting, I think the important distinction is between the binary finding libraries (this is the purpose of detect binary files with prefix) and finding config files (may default to conda installation correctly; if not, need to set environment var.)

To set environment vars,  see http://conda.pydata.org/docs/using/envs.html#saved-environment-variables

**Comment by kmuehlbauer** [comment]
@msarahan Thanks for clarification. This means, we do not need `detect_binary_files_with_prefix`? 

If so, we need to find the source where eccodes is setting those path's (its the conda build env path) and change it (which would be nicest) or add the path to the env as you suggested.

**Comment by kmuehlbauer** [comment]
@ocefpaf 

OK, setting `detect_binary_files_with_prefix` helped to fix the error in the test. So I think, linux is fine then. I can't see anything obvious why osx fails. It compiles but only a few tests are passing.

**Comment by ocefpaf** [comment]
> So I think, linux is fine then.

:+1: 

>  I can't see anything obvious why osx fails. It compiles but only a few tests are passing.

Maybe we need to implement something like this:

https://github.com/conda-forge/ecmwf_grib-feedstock/blob/master/recipe/build.sh

**Comment by kmuehlbauer** [comment]
OK, there the build-path/lib-folder is used as LIBRARY_SEARCH_PATH. Can't we just `make install` before `ctest`? Then, all libs should be at the right place, right?

**Comment by ocefpaf** [comment]
> OK, there the build-path/lib-folder is used as LIBRARY_SEARCH_PATH. Can't we just `make install` before `ctest`? Then, all libs should be at the right place, right?

Interesting question. When compiling something local I would say definitely no. However, in conda the package is created only if the tests passes, so that would be OK. @jakirkham any comments?

**Comment by msarahan** [comment]
Sounds fine to me. Just make sure it really doesn't create bad packages.

**Comment by kmuehlbauer** [comment]
Hmm, this didn't help. Switching back.

**Comment by kmuehlbauer** [comment]
JFTR: The passing tests are not really tests, but the download of the test datasets. Good to know, that this works, though. Also, defining that verbosity-switch didn't do anything.

**Comment by kmuehlbauer** [comment]
First tests passed on travis, @ocefpaf, thanks for the hint!

**Comment by kmuehlbauer** [comment]
No, one test is failing, 1 of 96. Bit lost now, how to debug this.

**Comment by ocefpaf** [comment]
@kmuehlbauer try reporting that back to the devs, maybe it is a known thing. If you cannot figure it out we can add a patch to remove that 1 test and address it later.

**Comment by kmuehlbauer** [comment]
One more try...

**Comment by kmuehlbauer** [comment]
Örks, @ocefpaf, how and where do we usually download accompanying tarballs? `wget` doesn't seem to work.

**Comment by ocefpaf** [comment]
Use `curl -O <url>`

**Comment by ocefpaf** [comment]
I wonder if we should make a data package... Maybe that is an overkill.

**Comment by kmuehlbauer** [comment]
Do I need to have curl in the requirements, then?

**Comment by ocefpaf** [comment]
> Do I need to have curl in the requirements, then?

Nope. That is built-in.

**Comment by kmuehlbauer** [comment]
Last resort: trying to output log-file of failing test. If this doesn't give us much, then we have at least something for the eccodes-developers.

**Comment by kmuehlbauer** [comment]
OK finally I got some evidence:

```
+ gts_ls data/gts/EGRR20150317121020_00493212.DAT
dyld: Library not loaded: libpng16.16.dylib
  Referenced from: /Users/travis/miniconda/envs/_build_placehold_placehold_placehold_placehold_plac/bin/gts_ls
  Reason: image not found
/Users/travis/build/conda-forge/staged-recipes/recipes/eccodes/build.sh: line 32:  5421 Trace/BPT trap: 5       gts_ls data/gts/EGRR20150317121020_00493212.DAT
Command failed: /bin/bash -x -e /Users/travis/build/conda-forge/staged-recipes/recipes/eccodes/build.sh
```
@ocefpaf
It seems, that something is wrong with libpng. Any ideas?

**Comment by ocefpaf** [comment]
> It seems, that something is wrong with libpng. Any ideas?

Not really. All the `libpng` available are from `defaults`. Maybe we should create our own version and check. (I will work on that soon, but note that the feedstock might not get created due to a problem we are facing with CircleCI. Hopefully that will be fixed soon.)

```
libpng                       1.2.50                        0  defaults        
                             1.5.13                        0  defaults        
                             1.5.13                        1  defaults        
                          .  1.6.17                        0  defaults
```

**Comment by kmuehlbauer** [comment]
We could disable `png` in call to cmake and see what happens. Then we can later work that out. I'll look into it after some :zzz:

**Comment by ocefpaf** [comment]
> We could disable `png` in call to cmake and see what happens. Then we can later work that out. I'll look into it after some :zzz: 

Go for it, but only for OS X. If things work and we are able to create the feedstock let's just open an issue to add `png` support for OS X.

File: recipes/eccodes/build.sh (hunk 5 above)
Comment (gillins) [inline recipes/eccodes/build.sh]: I'd start with seeing if you could run a similar cmake command on Windows. cmake seems pretty solid with different versions of VC etc. .vcproj files are usually for one version of VC, and can be fiddly to build from the command line.

**Comment by kmuehlbauer** [comment]
@ocefpaf I do not know, what to do next, which changes from #472 should I incorporate here? Or shall we wait until `libpng` is available within conda-forge. Builds are green there.

**Comment by kmuehlbauer** [comment]
Ahh, I think I finally got it. Cherry picked the libnpg recipe.

But the problem is still there.

**Comment by kmuehlbauer** [comment]
With the new libpng, we get this `Abort trap: 6` error with the failing command. Seems, there is something really wrong. Why does this only happen on OSX, not on linux?

```bash
+ gts_ls data/gts/EGRR20150317121020_00493212.DAT
data/gts/EGRR20150317121020_00493212.DAT
/Users/travis/build/conda-forge/staged-recipes/recipes/eccodes/build.sh: line 31: 10254 Abort trap: 6           gts_ls data/gts/EGRR20150317121020_00493212.DAT
Command failed: /bin/bash -x -e /Users/travis/build/conda-forge/staged-recipes/recipes/eccodes/build.sh

```

**Comment by kmuehlbauer** [comment]
Just for reference, this is how the problematic code is compiled and linked:

```bash
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f tools/CMakeFiles/gts_ls.dir/build.make tools/CMakeFiles/gts_ls.dir/depend
cd /Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build && /Users/travis/miniconda/envs/_build_placehold_placehold_placehold_placehold_plac/bin/cmake -E cmake_depends "Unix Makefiles" /Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source /Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/tools /Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build /Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build/tools /Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build/tools/CMakeFiles/gts_ls.dir/DependInfo.cmake --color=
Dependee "/Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build/tools/CMakeFiles/gts_ls.dir/DependInfo.cmake" is newer than depender "/Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build/tools/CMakeFiles/gts_ls.dir/depend.internal".
Dependee "/Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build/tools/CMakeFiles/CMakeDirectoryInformation.cmake" is newer than depender "/Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build/tools/CMakeFiles/gts_ls.dir/depend.internal".
Scanning dependencies of target gts_ls
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f tools/CMakeFiles/gts_ls.dir/build.make tools/CMakeFiles/gts_ls.dir/build
[ 71%] Building C object tools/CMakeFiles/gts_ls.dir/gts_ls.c.o
cd /Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build/tools && /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc  -DHAVE_ECCODES_CONFIG_H -I/Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build -I/Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/src -I/Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build/src -I/Users/travis/miniconda/envs/_build_placehold_placehold_placehold_placehold_plac/include  -arch x86_64 -fPIC -I/Users/travis/miniconda/envs/_build_placehold_placehold_placehold_placehold_plac/include  -O2 -g -DNDEBUG -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk -mmacosx-version-min=10.6 -fPIE   -o CMakeFiles/gts_ls.dir/gts_ls.c.o   -c /Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/tools/gts_ls.c
[ 71%] Linking C executable ../bin/gts_ls
cd /Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build/tools && /Users/travis/miniconda/envs/_build_placehold_placehold_placehold_placehold_plac/bin/cmake -E remove /Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build/bin/gts_ls
cd /Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build/tools && /Users/travis/miniconda/envs/_build_placehold_placehold_placehold_placehold_plac/bin/cmake -E cmake_link_script CMakeFiles/gts_ls.dir/link.txt --verbose=1
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc   -arch x86_64 -fPIC -I/Users/travis/miniconda/envs/_build_placehold_placehold_placehold_placehold_plac/include  -O2 -g -DNDEBUG -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk -mmacosx-version-min=10.6 -Wl,-search_paths_first -Wl,-headerpad_max_install_names   -arch x86_64 -L/Users/travis/miniconda/envs/_build_placehold_placehold_placehold_placehold_plac/lib -Wl,-rpath,/Users/travis/miniconda/envs/_build_placehold_placehold_placehold_placehold_plac/lib CMakeFiles/gts_ls.dir/gts_ls.c.o  -o ../bin/gts_ls  libgrib_tools.a ../lib/libeccodes.dylib /Users/travis/miniconda/envs/_build_placehold_placehold_placehold_placehold_plac/lib/libpng.dylib /Users/travis/miniconda/envs/_build_placehold_placehold_placehold_placehold_plac/lib/libz.dylib /Users/travis/miniconda/envs/_build_placehold_placehold_placehold_placehold_plac/lib/libnetcdf.dylib /Users/travis/miniconda/envs/_build_placehold_placehold_placehold_placehold_plac/lib/libhdf5.dylib /Users/travis/miniconda/envs/_build_placehold_placehold_placehold_placehold_plac/lib/libhdf5_hl.dylib /usr/lib/libdl.dylib /usr/lib/libm.dylib /Users/travis/miniconda/envs/_build_placehold_placehold_placehold_placehold_plac/lib/libjasper.dylib /Users/travis/miniconda/envs/_build_placehold_placehold_placehold_placehold_plac/lib/libjpeg.dylib -Wl,-rpath,@loader_path/../lib 
[ 71%] Built target gts_ls
```

**Comment by ocefpaf** [comment]
> Ahh, I think I finally got it. Cherry picked the libnpg recipe.

Cool. Sorry for not being more specific before. (But now that #472 got merged we have to remove it from here once the feedstock is working.)

> But the problem is still there.

```
+ gts_ls data/gts/EGRR20150317121020_00493212.DAT
data/gts/EGRR20150317121020_00493212.DAT
/Users/travis/build/conda-forge/staged-recipes/recipes/eccodes/build.sh: line 31: 10254 Abort trap: 6           gts_ls data/gts/EGRR20150317121020_00493212.DAT
Command failed: /bin/bash -x -e /Users/travis/build/conda-forge/staged-recipes/recipes/eccodes/build.sh
```

I no longer see the `dyld: Library not loaded: libpng16.16.dylib` so there is some progress :smile: 

@jakirkham you are the only OS X expert I know :wink: can you take a look at this once you have some free time?

**Comment by kmuehlbauer** [comment]
OK here is the verbose ctest output for reference:

```ctest
test 18
      Start 18: eccodes_t_gts_ls

18: Test command: /Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/tests/gts_ls.sh
18: Test timeout computed to be: 1500
18: +++ proj_dir=/Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source
18: +++ data_dir=/Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build/data
18: +++ def_dir=/Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build/share/eccodes/definitions
18: +++ ECCODES_DEFINITION_PATH=/Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build/share/eccodes/definitions
18: +++ export ECCODES_DEFINITION_PATH
18: +++ tools_dir=/Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build/bin/
18: +++ tigge_dir=/Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build/bin/
18: +++ test_dir=/Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build/tests
18: +++ samp_dir=/Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build/share/eccodes/samples
18: +++ ECCODES_SAMPLES_PATH=/Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build/share/eccodes/samples
18: +++ export ECCODES_SAMPLES_PATH
18: + cd /Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build/data/gts
18: + label=gts_ls_test
18: + fLog=gts_ls_test.log
18: + rm -f gts_ls_test.log
18: + touch gts_ls_test.log
18: + fTmp=gts_ls_test.tmp.txt
18: + rm -f gts_ls_test.tmp.txt
18: + gts_file=EGRR20150317121020_00493212.DAT
18: + f=EGRR20150317121020_00493212.DAT
18: + echo EGRR20150317121020_00493212.DAT
18: + /Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/build/bin//gts_ls EGRR20150317121020_00493212.DAT
18: /Users/travis/miniconda/conda-bld/work/eccodes-0.13.0-Source/tests/gts_ls.sh: line 35: 16513 Abort trap: 6           ${tools_dir}/gts_ls $f >> $fLog
23/96 Test #18: eccodes_t_gts_ls ....................***Failed    1.50 sec
```

**Comment by kmuehlbauer** [comment]
How do we proceed? I'll inform the eccodes-dev about that issue. Shall we leave out that specific test?

File: recipes/eccodes/meta.yaml (hunk 4 above)
Comment (pelson) [inline recipes/eccodes/meta.yaml]: Yes, please feel free to add me to this.

**Comment by pelson** [comment]
This was a long read :smile: 

> How do we proceed? I'll inform the eccodes-dev about that issue. Shall we leave out that specific test?

I've just ran the build on my own mac and am getting lots of ```Trace/BPT trap: 5``` type messages. I'll try digging a little deeper, but I definitely think it is worth getting some input from the eccodes devs.

**Comment by pelson** [comment]
Would be lovely if conda-build had a way of expressing this:

```
Name clashes prevent a simultaneous use of the GRIB-API and ecCodes libraries.
```

**Comment by pelson** [comment]
Yay to WMO tables... 😄 

```number of files: 11721```

**Comment by kmuehlbauer** [comment]
> Would be lovely if conda-build had a way of expressing this:
> 
> Name clashes prevent a simultaneous use of the GRIB-API and ecCodes libraries.

They will abandon GRIB-API in the near future in favour of ecCodes. ecCodes will directly start with version 2.0 to avoid name clashes with the GRIB-API

**Comment by ocefpaf** [comment]
```number of files: 11721```

:scream:

**Comment by kmuehlbauer** [comment]
I've sent an email with a short description and the build-logs (both linux and macosx). to the ecCodes-devs.

**Comment by ocefpaf** [comment]
> I've sent an email with a short description and the build-logs (both linux and macosx). to the ecCodes-devs.

Thanks for doing this @kmuehlbauer. Some builds are very painful, but I believe it will be worth in the end.

**Comment by kmuehlbauer** [comment]
I made a force-reset to the changes before the cherry pick, because `libpng` is now available within conda-forge.

**Comment by kmuehlbauer** [comment]
ECMWF suggests not to build inplace. So trying that.

**Comment by kmuehlbauer** [comment]
Note to self: read things twice :see_no_evil:

**Comment by kmuehlbauer** [comment]
> Note to self: read things twice :see_no_evil:

This is certainly not my day...

**Comment by ocefpaf** [comment]
@kmuehlbauer since you are skipping Windows this is good, right?

File: recipes/eccodes/build.sh [hunk 6]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env bash
+
+if [[ $(uname) == Darwin ]]; then
+  export LIBRARY_SEARCH_VAR=DYLD_FALLBACK_LIBRARY_PATH
+  ctestarg="-E gts_ls"
+elif [[ $(uname) == Linux ]]; then
+  export LIBRARY_SEARCH_VAR=LD_LIBRARY_PATH
+fi
+
+export PYTHON=
```
Comment (ocefpaf) [inline recipes/eccodes/build.sh]: I think that line 23 should take care of this, no?

**Pull request merged**